File: import_excel_to_api.py
import os
from datetime import datetime
from pathlib import Path

import pandas as pd
import requests

BASE_DIR = Path(__file__).resolve().parent
EXCEL_CANDIDATES = sorted(BASE_DIR.glob("DataServiceDR*.xlsx"))
EXCEL_FILE = str(EXCEL_CANDIDATES[0]) if EXCEL_CANDIDATES else str(BASE_DIR / "DataServiceDR.xlsx")
API_URL = os.getenv("API_URL", "http://localhost:4001/api/requests")

FIELD_ALIASES = {
    "customer": ["ลูกค้า", "Customer", "Customer Name"],
    "ref": ["Ref.", "Ref", "ref", "ระดับ", "Level"],
    "source": ["ช่องทาง", "Source", "Channel"],
    "receivedAt": ["วันเวลารับแจ้ง", "Received At", "Date Received", "วันที่รับแจ้ง"],
    "ticket": ["เลขติดตาม", "Ticket", "Tracking Number", "เลขติดตามงาน"],
    "location": ["สถานที่", "Location", "Site", "สถานที่ติดตั้ง"],
    "contact": ["ผู้ติดต่อ", "Contact", "Contact Name"],
    "phone": ["Phone", "เบอร์โทร", "โทรศัพท์"],
    "description": ["รายละเอียด", "Description", "Issue Detail"],
    "jobType": ["ลักษณะงาน", "Job Type", "ประเภทงาน"],
    "status": ["สถานะ", "Status"],
    "assignee": ["ผู้ดำเนินการ", "Assignee", "Technician"],
    "appointment": ["วันเวลานัดหมาย", "Appointment", "Scheduled Time"],
    "completedAt": ["วันเวลาเสร็จ", "Completed At", "Finished At"],
    "action": ["การดำเนินการ", "Action", "Work Done"],
    "notes": ["หมายเหตุ", "Notes", "Remark"],
}


def normalize_string(value):
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return ""
    return str(value).strip()


def normalize_datetime(value):
    if value is None or (isinstance(value, float) and pd.isna(value)):
        return None
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        try:
            return datetime.fromisoformat(text.replace("Z", "+00:00")).isoformat()
        except Exception:
            try:
                return pd.to_datetime(text).isoformat()
            except Exception:
                return None
    if hasattr(value, "isoformat"):
        try:
            return value.isoformat()
        except Exception:
            return None
    try:
        return pd.to_datetime(value).isoformat()
    except Exception:
        return None


def get_value(row, names):
    for name in names:
        if name in row and not pd.isna(row.get(name)):
            return row.get(name)
    return None


def build_payload(row):
    payload = {}
    for field, aliases in FIELD_ALIASES.items():
        value = get_value(row, aliases)
        if field in {"receivedAt", "appointment", "completedAt"}:
            payload[field] = normalize_datetime(value)
        else:
            payload[field] = normalize_string(value)
    return payload


def load_excel_rows(path):
    xls = pd.ExcelFile(path)
    all_rows = []

    for sheet in xls.sheet_names:
        try:
            df = pd.read_excel(path, sheet_name=sheet)
        except Exception as exc:
            print(f"Skip sheet '{sheet}' due to read error: {exc}")
            continue

        if df.empty:
            continue

        for _, row in df.iterrows():
            normalized_row = row.to_dict()
            all_rows.append(build_payload(normalized_row))

    return all_rows


def main():
    if not os.path.exists(EXCEL_FILE):
        raise FileNotFoundError(f"ไม่พบไฟล์ Excel: {EXCEL_FILE}")

    print(f"กำลังอ่านไฟล์: {EXCEL_FILE}")
    rows = load_excel_rows(EXCEL_FILE)
    print(f"พบข้อมูลทั้งหมด {len(rows)} rows")

    success = 0
    failed = 0
    for index, payload in enumerate(rows, start=1):
        try:
            response = requests.post(API_URL, json=payload, timeout=30)
            if response.status_code >= 400:
                print(f"Row {index} failed: status={response.status_code} payload={payload}")
                print(response.text)
                failed += 1
                continue
            success += 1
            if index % 50 == 0:
                print(f"ส่งแล้ว {index}/{len(rows)} rows")
        except Exception as exc:
            failed += 1
            print(f"Row {index} exception: {exc}")
            print(payload)

    print(f"Import complete: success={success}, failed={failed}")


if __name__ == "__main__":
    main()

File: import_excel_to_api.mjs
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import xlsx from 'xlsx';

const XLSX = xlsx;

const root = path.dirname(fileURLToPath(import.meta.url));
const candidates = fs.readdirSync(root)
  .filter((name) => /^DataServiceDR.*\.xlsx$/i.test(name))
  .sort();
const workbookPath = path.join(root, candidates[0] || 'DataServiceDR.xlsx');
const apiUrl = `${process.env.API_URL || 'http://localhost:4001'}/api/requests`;

const aliases = {
  customer: ['ลูกค้า', 'Customer', 'Customer Name'],
  ref: ['Ref.', 'Ref', 'ref', 'ระดับ', 'Level'],
  source: ['ช่องทาง', 'แหล่งที่มา', 'Source', 'Channel'],
  receivedAt: ['วันเวลารับแจ้ง', 'วันเวลาที่รับแจ้ง', 'Received At', 'Date Received', 'วันที่รับแจ้ง'],
  ticket: ['เลขติดตาม', 'เลขที่ติดตามงาน', 'Ticket', 'Tracking Number', 'เลขติดตามงาน'],
  location: ['สถานที่', 'สถานที่/สาขา', 'Location', 'Site', 'สถานที่ติดตั้ง'],
  site: ['สถานที่ตั้ง', 'Site Address'],
  contact: ['ผู้ติดต่อ', 'Contact', 'Contact Name'],
  phone: ['Phone', 'เบอร์ติดต่อ', 'เบอร์โทร', 'โทรศัพท์'],
  description: ['รายละเอียด', 'ข้อมูลการรับแจ้ง', 'Description', 'Issue Detail'],
  image: ['รูปภาพที่แจ้ง', 'Image'],
  ma: ['MA'],
  jobType: ['ลักษณะงาน', 'Job Type', 'ประเภทงาน'],
  status: ['สถานะ', 'สถานะงาน', 'Status'],
  assignee: ['ผู้ดำเนินการ', 'Assignee', 'Technician'],
  appointment: ['วันเวลานัดหมาย', 'วันที่นัดหมาย เวลาเริ่มต้น', 'Appointment', 'Scheduled Time'],
  appointmentEnd: ['วันที่นัดหมาย เวลาสิ้นสุด', 'Appointment End'],
  action: ['การดำเนินการ', 'Action', 'Work Done'],
  result: ['ผลการดำเนินการ', 'Result'],
  equipment: ['เกี่ยวกับอุปกรณ์', 'Equipment'],
  completedImage: ['รูปภาพที่ดำเนินการเสร็จแล้ว', 'Completed Image'],
  completedAt: ['วันเวลาเสร็จ', 'วันเวลาเสร็จ', 'Completed At', 'Finished At'],
  map: ['MAP', 'Map'],
  vehicle: ['ทะเบียนรถ', 'Vehicle'],
  notes: ['หมายเหตุ', 'Notes', 'Remark'],
  file: ['ไฟล์', 'File'],
};

const dateFields = new Set(['receivedAt', 'appointment', 'appointmentEnd', 'completedAt']);
const latestSourceDate = Date.parse(process.env.LATEST_SOURCE_DATE || '2026-08-28T23:59:59+07:00');
const cleanHeader = (value) => String(value ?? '').trim();
const cleanString = (value) => value == null ? '' : String(value).trim();
const normalizeYear = (date) => {
  if (date.getUTCFullYear() >= 2400) {
    date.setUTCFullYear(date.getUTCFullYear() - 543);
  }
  if (date.getTime() > latestSourceDate) {
    date.setUTCFullYear(date.getUTCFullYear() - 1);
  }
  return date.toISOString();
};
const dateValue = (value) => {
  if (value == null || value === '') return null;
  if (value instanceof Date) return normalizeYear(new Date(value));
  if (typeof value === 'number') {
    const parsed = XLSX.SSF.parse_date_code(value);
    if (!parsed) return null;
    return normalizeYear(new Date(Date.UTC(parsed.y, parsed.m - 1, parsed.d, parsed.H || 0, parsed.M || 0, parsed.S || 0)));
  }
  const text = String(value).trim();
  const parsed = new Date(text);
  return Number.isNaN(parsed.getTime()) ? null : normalizeYear(parsed);
};

const valueFor = (row, names) => {
  const entries = Object.entries(row);
  const match = entries.find(([key, value]) => names.includes(cleanHeader(key)) && value != null && value !== '');
  return match?.[1];
};

const payloadFor = (row) => Object.fromEntries(Object.entries(aliases).map(([field, names]) => [
  field,
  dateFields.has(field) ? dateValue(valueFor(row, names)) : cleanString(valueFor(row, names)),
]));

if (!fs.existsSync(workbookPath)) throw new Error(`ไม่พบไฟล์ Excel: ${workbookPath}`);
const workbook = XLSX.readFile(workbookPath, { cellDates: false });
const rows = workbook.SheetNames.flatMap((sheetName) => {
  const sheet = workbook.Sheets[sheetName];
  return XLSX.utils.sheet_to_json(sheet, { defval: '' }).map(payloadFor);
});

console.log(`กำลังอ่านไฟล์: ${workbookPath}`);
console.log(`พบข้อมูลทั้งหมด ${rows.length} rows จาก ${workbook.SheetNames.length} sheets`);

const existingResponse = await fetch(apiUrl);
if (!existingResponse.ok) throw new Error(`อ่านข้อมูลเดิมไม่สำเร็จ (${existingResponse.status})`);
const existingRows = await existingResponse.json();
const rowKey = (row) => row.ticket
  ? `ticket:${row.ticket}`
  : `fallback:${row.customer}|${row.completedAt || ''}|${row.assignee || ''}|${row.notes || ''}`;
const existingByKey = new Map(existingRows.map((row) => [rowKey(row), row]));
const dataRows = rows.filter((row) => row.customer || row.ticket || row.description);
const pendingRows = dataRows.filter((row) => !existingByKey.has(rowKey(row)));
const updateRows = dataRows.filter((row) => existingByKey.has(rowKey(row)));
console.log(`พบข้อมูลเดิมสำหรับอัปเดต ${updateRows.length} rows`);
console.log(`เหลือข้อมูลใหม่ที่ต้องนำเข้า ${pendingRows.length} rows`);

let success = 0;
let failed = 0;
const importRow = async (payload, index) => {
  try {
    const existing = existingByKey.get(rowKey(payload));
    const response = await fetch(existing ? `${apiUrl}/${existing.id}` : apiUrl, {
      method: existing ? 'PUT' : 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
    });
    if (!response.ok) {
      failed += 1;
      console.error(`Row ${index + 1} failed (${response.status}): ${await response.text()}`);
      return;
    }
    success += 1;
  } catch (error) {
    failed += 1;
    console.error(`Row ${index + 1} exception: ${error.message}`);
  }
};

for (let start = 0; start < dataRows.length; start += 8) {
  const batch = dataRows.slice(start, start + 8);
  await Promise.all(batch.map((payload, offset) => importRow(payload, start + offset)));
  console.log(`ส่งแล้ว ${Math.min(start + batch.length, dataRows.length)}/${dataRows.length} rows`);
}

console.log(`Import complete: success=${success}, failed=${failed}`);
if (failed > 0) process.exitCode = 1;
